### PR TITLE
Added Node and fabric data to BDXOtaSender to keep track of connections

### LINKED ISSUE #16044: OTA Provider cannot handle multiple QueryImage with no downloads in between
#### Problem
The OTA Provider is not cleaning up a BDX transfer session when the requestor does not proceed with an image download but another QueryImage is received.

Steps to reproduce:
1. Run ota-provider-app with the --filepath option
2. Initiate QueryImage from ota-requestor-app (via announce-ota-provider command or periodic query)
3. After image successfully downloaded, initiate another QueryImage
4. After requestor receives the QueryImageResponse, the image supplied will be ignored since the software version matches
5. Initiate another QueryImage -> the ota-provider-app will report an error about initializing BDX transfer session (see logs below):

```
[1646867628654] [23847:260368] CHIP: [DMG] Received command for Endpoint=0 Cluster=0x0000_0029 Command=0x0000_0000
[1646867628654] [23847:260368] CHIP: [ZCL] OTA Provider received QueryImage
[1646867628654] [23847:260368] CHIP: [ZCL]   VendorID: 0xfff1
[1646867628654] [23847:260368] CHIP: [ZCL]   ProductID: 32769
[1646867628654] [23847:260368] CHIP: [ZCL]   SoftwareVersion: 0
[1646867628654] [23847:260368] CHIP: [ZCL]   ProtocolsSupported: [
[1646867628655] [23847:260368] CHIP: [ZCL]     0
[1646867628655] [23847:260368] CHIP: [ZCL]   ]
[1646867628655] [23847:260368] CHIP: [ZCL]   HardwareVersion: 0
[1646867628655] [23847:260368] CHIP: [ZCL]   Location: XX
[1646867628655] [23847:260368] CHIP: [ZCL]   RequestorCanConsent: 1
[1646867628655] [23847:260368] CHIP: [SWU] generated updateToken: 9493EB4A5719A009E4A779FC9CA49C
[1646867628655] [23847:260368] CHIP: [DIS] Fabric (1) loaded from storage. Calling OnFabricRetrievedFromStorage
[1646867628655] [23847:260368] CHIP: [DIS] Fabric (1) loaded from storage. Calling OnFabricRetrievedFromStorage
[1646867628655] [23847:260368] CHIP: [ZCL] OpCreds: Fabric 0x1 was retrieved from storage. FabricId 0x0000000000000001, NodeId 0x0000000000C0FFEE, VendorId 0xFFF1
[1646867628655] [23847:260368] CHIP: [ZCL] OpCreds: Call to fabricListChanged
[1646867628655] [23847:260368] CHIP: [DMG] Endpoint 0, Cluster 0x0000_003E update version to d24641b7
[1646867628655] [23847:260368] CHIP: [DMG] Endpoint 0, Cluster 0x0000_003E update version to d24641b8
[1646867628655] [23847:260368] CHIP: [SWU] Generated URI: bdx://0000000000C0FFEE/tmp/test.bin
[1646867628655] [23847:260368] CHIP: [BDX] Failed to initialize BDX transfer session: ../../examples/ota-provider-app/linux/third_party/connectedhomeip/src/protocols/bdx/BdxTransferSession.cpp:182: CHIP Error 0x00000003: Incorrect state
```

 #### Proposed Solution
Either we should not call `mBdxOtaSender.PrepareForTransfer` until we are actually starting the transfer or allow multiple calls to this function.

### PR DESCRIPTION
**Pull request by harsha-rajendran**
#### Problem
What is being fixed?  #16044 

#### Change overview
The BDXOtaSender keeps track of the fabricIdx and nodeId of the "active" requestor. If the requestor hangs/dies and re-attempts a connection, the previous BDX session will be cleaned up, to allow the connection. If a new node tries to connect in the interim a BUSY message will be sent to it, requesting it to retry later. Note that the BDX connection will eventually timeout on no-activity and should reset automatically as well.

#### Testing
- Tested the following scenarios:
- If a requestor fails to continue an OTA after getting a QueryImage response (i.e. crashes), a new Query-image can be sent immediately, upon re-connecting.
- A requestor does not continue an OTA after getting a QueryImage response (i.e. crashes). If a new node sends a queryImage to the provider, it will get a BUSY response.

### CODE REVIEW

File: examples/ota-provider-app/ota-provider-common/BdxOtaSender.cpp [hunk 1]
```diff
@@ -35,6 +35,32 @@ BdxOtaSender::BdxOtaSender()
...
+        // Reset stale connection from the Same Node if exists
+        if ((mFabricIndex.HasValue() && mFabricIndex.Value() == fabricIndex) && (mNodeId.HasValue() && mNodeId.Value() == nodeId))
+        {
+            Reset();
+        }
+        // Prevent a new node connection since another is active
+        else if ((mFabricIndex.HasValue() && mFabricIndex.Value() != fabricIndex) ||
+                 (mNodeId.HasValue() && mNodeId.Value() != nodeId))
+        {
+            return CHIP_ERROR_INTERNAL; // busy
```
Comment (tcarmelveilleux) [inline examples/ota-provider-app/ota-provider-common/BdxOtaSender.cpp:51]: Why not create a new CHIP_ERROR_BUSY ?

Comment (hrajendran-samsung) [reply examples/ota-provider-app/ota-provider-common/BdxOtaSender.cpp:51]: Seems prudent!

File: examples/ota-provider-app/esp32/main/BdxOtaSender.cpp [hunk 2]
```diff
@@ -28,6 +28,32 @@ using chip::bdx::StatusCode;
 using chip::bdx::TransferControlFlags;
 using chip::bdx::TransferSession;
 
+CHIP_ERROR BdxOtaSender::InitializeTransfer(chip::FabricIndex fabricIndex, chip::NodeId nodeId)
```
Comment (hrajendran-samsung) [inline examples/ota-provider-app/esp32/main/BdxOtaSender.cpp:31]: Cleaner to have one BDXOtaSender class, instead of two very similar ones. Ticket for the backlog, perhaps and maybe someone from Espressif can shed more light here?

Comment (carol-apple) [reply examples/ota-provider-app/esp32/main/BdxOtaSender.cpp:31]: Ideally, we had refactored the OTA Provider in a similar fashion as OTA Requestor which could allow platform specific impl to be done with most code being common. This couldn't be achieved in the v1 timeframe so for the time being, the ota-provider-app for each platform has to make a pretty similar copy.

File: src/lib/core/CHIPError.h [hunk 3]
```diff
@@ -2400,6 +2400,15 @@ using CHIP_ERROR = ::chip::ChipError;
  */
 #define CHIP_ERROR_INVALID_FILE_IDENTIFIER                     CHIP_CORE_ERROR(0xda)
 
+/**
+ * @def CHIP_ERROR_BUSY
+ *
+ * @brief
+ *   The Resource is busy and cannot process the request. Trying again might work.
+ */
+#define CHIP_ERROR_BUSY                     CHIP_CORE_ERROR(0xdb)
```
Comment (carol-apple) [inline src/lib/core/CHIPError.h:2409]: I'm surprised restyle doesn't catch this!

File: examples/ota-provider-app/ota-provider-common/BdxOtaSender.cpp [hunk 4]
```diff
@@ -35,6 +35,32 @@ BdxOtaSender::BdxOtaSender()
...
+{
+    if (mInitialized)
+    {
+        // Reset stale connection from the Same Node if exists
+        if ((mFabricIndex.HasValue() && mFabricIndex.Value() == fabricIndex) && (mNodeId.HasValue() && mNodeId.Value() == nodeId))
+        {
+            Reset();
+        }
+        // Prevent a new node connection since another is active
+        else if ((mFabricIndex.HasValue() && mFabricIndex.Value() != fabricIndex) ||
```
Comment (carol-apple) [inline examples/ota-provider-app/ota-provider-common/BdxOtaSender.cpp:48]: Do `mFabricIndex` and `mNodeId` need to be optional? We could just initialize/reset to kUndefinedFabricIndex/kUndefinedNodeId and makes the check here a little easier?

**Review by carol-apple — APPROVED** [review]

**Review by selissia — APPROVED** [review]

File: examples/ota-provider-app/ota-provider-common/BdxOtaSender.cpp (hunk 4 above)
Comment (hrajendran-samsung) [reply examples/ota-provider-app/ota-provider-common/BdxOtaSender.cpp:48]: True! We could use an invalid value. But I guess the whole point of using an "optional" was so we don't have to set aside an invalid value. Luckily in this case the spec already defined invalid values (unless it changes in v2.0 ?).
TLDR; I am fine either way, unless you have a strong preference. :-)

File: src/lib/core/CHIPError.h (hunk 3 above)
Comment (hrajendran-samsung) [reply src/lib/core/CHIPError.h:2409]: All the other error codes seem to be following a similar spacing paradigm...

Comment (carol-apple) [reply src/lib/core/CHIPError.h:2409]: It seems like all the other `CHIP_CORE_ERROR` line up in the file...

File: examples/ota-provider-app/ota-provider-common/BdxOtaSender.cpp (hunk 4 above)
Comment (carol-apple) [reply examples/ota-provider-app/ota-provider-common/BdxOtaSender.cpp:48]: I don't feel strongly about it, just thought you could have less checks here for `HasValue()`

File: src/lib/core/CHIPError.h (hunk 3 above)
Comment (hrajendran-samsung) [reply src/lib/core/CHIPError.h:2409]: All the error codes seem to line up, until CHIP_ERROR_IM_MALFORMED_COMMAND_STATUS_IB (line 2233). From there on it gets wonky.
